### PR TITLE
Phase 3 R10: Spectral Fourier Loss for Surface Pressure (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The standard L1 loss treats all spatial errors equally, but tandem-foil pressure distributions have higher-frequency variations (wake interference, vortex impingement, multi-scale structures). A spectral loss operating on the 1D Fourier transform of surface pressure predictions could directly penalize high-frequency errors that L1 underweights.

The Fourier Amplitude and Correlation Loss (FACL, NeurIPS 2024) showed that supplementing L2 loss with spectral losses consistently improves high-frequency detail recovery in weather prediction — which faces a similar "blurring" problem to our surface pressure prediction.

This is Lion-compatible since it only modifies the loss computation, not the architecture.

## Instructions
Pull latest noam. SENPAI_TIMEOUT_MINUTES=180, SENPAI_MAX_EPOCHS=500. Use `--wandb_group "phase3-r10-spectral"`.

### Implementation
For surface nodes only:
1. Sort surface nodes by x-coordinate (arc-length proxy) for each foil separately
2. Compute 1D FFT of predicted and target surface pressure along the sorted surface
3. Compute spectral loss: `|FFT(pred_p) - FFT(target_p)|` in amplitude space
4. Weight higher frequencies more heavily (linear ramp: weight_k = k/K)
5. Add as `spectral_weight * spectral_loss` to main loss

```python
# Pseudo-code for spectral loss:
for b in range(B):
    surf_idx = surf_mask[b].nonzero().squeeze()
    if surf_idx.numel() < 16: continue
    # Sort by x-coordinate
    xy_surf = xy[b, surf_idx]
    order = xy_surf[:, 0].argsort()
    pred_surf_p = pred[b, surf_idx[order], 2]  # pressure channel
    tgt_surf_p = target[b, surf_idx[order], 2]
    # FFT
    pred_fft = torch.fft.rfft(pred_surf_p)
    tgt_fft = torch.fft.rfft(tgt_surf_p)
    # Amplitude loss with frequency weighting
    K = pred_fft.shape[0]
    freq_weights = torch.linspace(0, 1, K, device=pred_fft.device)
    spectral_loss += (freq_weights * (pred_fft.abs() - tgt_fft.abs()).abs()).mean()
```

### GPU 0: Spectral loss, weight=0.1
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --wandb_name "edward/r10-spec-01" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 1: Spectral loss, weight=0.3
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --wandb_name "edward/r10-spec-03" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 2: Spectral loss, weight=0.5
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --wandb_name "edward/r10-spec-05" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 3: Spectral loss, weight=1.0 (equal to main loss)
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --wandb_name "edward/r10-spec-10" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 4: Spectral loss w=0.3, tandem-only (only apply to tandem samples)
```bash
CUDA_VISIBLE_DEVICES=4 python train.py --wandb_name "edward/r10-spec-tan-only" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 5: Spectral loss w=0.3 + DomainLayerNorm compound
```bash
CUDA_VISIBLE_DEVICES=5 python train.py --wandb_name "edward/r10-spec-dln" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 6: Spectral loss w=0.3, log-frequency weighting (log(k+1)/log(K+1))
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --wandb_name "edward/r10-spec-logfreq" --wandb_group "phase3-r10-spectral" --agent edward
```

### GPU 7: Focal Frequency Loss variant (penalize frequency bands with highest error)
Instead of fixed weighting, compute per-frequency error and upweight the worst bands.
```bash
CUDA_VISIBLE_DEVICES=7 python train.py --wandb_name "edward/r10-focal-freq" --wandb_group "phase3-r10-spectral" --agent edward
```

## Baseline
| val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|----------|------|--------|-------|------|---------||
| **0.3997** | ~13.8 | 8.8 | 33.2 | 24.8 | merged noam baseline |

---

## Results

All 8 variants ran for 180 min / ~244–246 epochs.

| Run | val/loss | p_in | p_oodc | p_tan | p_re | Ux_in | Uy_in | W&B Run |
|-----|----------|------|--------|-------|------|-------|-------|---------|
| Baseline | **0.3997** | ~13.8 | 8.8 | 33.2 | 24.8 | — | — | merged noam |
| spec-01 (w=0.1) | 0.7269 | 14.97 | 10.53 | 36.97 | 25.48 | 2.28 | 0.78 | anks4bca |
| spec-03 (w=0.3) | 0.7640 | 15.41 | 11.58 | 37.30 | 26.39 | 2.94 | 1.03 | afhcg43n |
| spec-05 (w=0.5) | 0.7804 | 15.53 | 12.03 | 37.56 | 26.42 | 4.69 | 1.11 | y6t9j6yh |
| spec-10 (w=1.0) | **7.2458** | 17.74 | **376.23** | 38.19 | **291.78** | 5.10 | 1.19 | zx3tv22w |
| spec-tan-only (w=0.3) | 0.7547 | 15.32 | 11.18 | 37.17 | 25.85 | 3.42 | 0.80 | 1y9kusnv |
| spec-dln (w=0.3+DLN) | 0.7508 | 16.00 | 11.22 | 36.60 | 25.94 | 3.24 | 0.89 | rov3hq76 |
| spec-logfreq (w=0.3,log) | 0.7628 | **14.17** | 11.80 | 37.36 | 26.08 | 3.57 | 0.95 | 78jf4lh0 |
| focal-freq (w=0.3,focal) | 0.8593 | 16.78 | 13.09 | 40.56 | 27.00 | 5.04 | 1.46 | ya7hm9bm |

Peak memory: ~36 GB (spec-01 GPU 0 snapshot). Spectral FFT adds negligible overhead vs standard training.

### What happened

**The spectral loss hurts on all variants.** Every configuration is worse than baseline across all metrics. Key patterns:

1. **Monotonic degradation with weight**: w=0.1 < 0.3 < 0.5, and w=1.0 diverges catastrophically on OOD splits (p_oodc=376, p_re=292). The spectral amplitude loss completely overwhelms the spatial L1 at weight=1.0.

2. **No variant beats baseline on any primary metric.** Even the best case (spec-01 p_oodc=10.53) is 20% worse than baseline 8.8.

3. **Log-freq weighting gets closest on p_in** (14.17 vs baseline 13.8) but regresses on all OOD splits.

4. **Focal frequency is the worst functional variant.** Adaptive per-frequency upweighting amplifies bad frequency bins mid-training, adding noise to gradients.

5. **DLN compound doesn't rescue the approach.** The DLN helps tandem (p_tan=36.60, best of the w=0.3 cluster) but doesn't offset the spectral overhead.

**Why it failed:**

- The FFT assumes uniform sampling, but surface nodes sorted by x-coordinate are not uniformly spaced in arc-length. The spectral representation is therefore physically meaningless — it's the FFT of an irregularly sampled signal treated as uniform, which introduces spectral leakage and aliasing.
- For tandem configs, the x-sort mixes nodes from both foils (which overlap in x), creating a discontinuous signal with large spurious high-frequency components. The spectral loss then penalizes these artifacts rather than genuine high-frequency pressure features.
- The known limitation `SURFACE_IDS=(5,6) misses boundary ID 7` means ~half of tandem surface nodes are absent from `surf_mask`, further corrupting the sorted signal for tandem samples.
- The linear frequency ramp (`weight_k = k/K`) starts at 0 for DC and ramps to 1 for Nyquist. This effectively de-emphasizes the mean pressure error (DC component) while heavily penalizing high-frequency noise. This is backwards for our use case: the mean pressure level and low-frequency shape (camber-induced pressure distribution) matter more than fine texture.

### Suggested follow-ups

- **Arc-length parameterization**: Sort by arc-length from leading edge rather than x-coordinate. This creates a meaningful uniform parameterization. Would require computing arc-length on-the-fly from node positions.
- **Per-foil FFT**: Separate surface nodes by boundary ID (5 vs 6) before computing FFT, avoiding tandem mixing. This would also avoid the ID-7 missing-surface problem for tandem.
- **Low-frequency emphasis**: Invert the frequency weighting (penalize DC and low-k errors more than high-k). The pressure distribution's macro shape (DC + fundamental) is more important than high-frequency oscillations.
- **Phase-aware loss**: Use complex FFT difference `|FFT(pred) - FFT(target)|` directly (both amplitude AND phase), not just amplitude difference. The current amplitude-only loss misses phase errors that shift pressure peaks.
- **Spectral regularization on Cp**: Apply to Cp = (p - p_ref) / q rather than raw pressure, making the spectral loss scale-invariant across Re numbers. This might explain why p_re (OOD Reynolds) is the worst-hurt metric.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing — spectral loss experiments all significantly worse than baseline.
